docs/DesignDocs/enclave_openssl_support.md

```diff
@@ -0,0 +1,90 @@
+
+- [ ] Add ctr_drbg.c to the `oecryptoopenssl` library implementation for parity with `oecryptombedtls`.
+- [ ] Update OpenSSL code changes to use `oe_get_entropy()` for entropy pool seeding, and avoid modifying the OpenSSL drbg code.
+- [ ] Formalize additional syscall stubs needed for OpenSSL as needed for [ssl_stubs.c](https://github.com/openenclave/openenclave-openssl/blob/master/0001-Get-openssl-to-build-against-MUSL-headers.-Use-SGX-r.patch).
+- [ ] Update the OpenSSL target branch to the latest [OpenSSL_1_1_1-stable](https://github.com/openssl/openssl/tree/OpenSSL_1_1_1-stable) and resolve test coverage differences or breaking changes.
+- [ ] Evaluate the set of [broken OpenSSL tests](https://github.com/openenclave/openenclave-openssl/blob/master/test/tests.broken) when executed in an enclave.
+
+## Alternates
+
+As mentioned in the motivation, there is a desire for a long-term solution to enable OE SDK users to replace the crypto library used by the OE enclave runtime with an arbitrary crypto library. One approach to enable this is to formalize the internal OE crypto API that is being implemented by `oecryptombedtls` and `oecryptoopenssl` in this proposal. This approach is an additive change that we may pursue in the future and does not obviate the support proposed here.
+
+An additional refinement of that approach is to eliminate the internal OE crypto API layer entirely:
+
+- OE SDK calls OpenSSL APIs directly instead of using its `oecrypto` indirection.
+- `oecryptoopenssl` becomes unnecessary, and `oecryptombedtls` would need to be refactored to be an mbedTLS to OpenSSL API adaptor instead.
+- Other crypto libraries that are already compatible or have existing shims with the OpenSSL API can be reused with minimal change.
+
+This option is not being pursued currently because:
+
+- As with the option to expose the existing `oecrypto` API for adaptation, it is additive to the proposed solution and can be incrementally pursued at a later date.
+  - Since it doesn't formalize the `oecrypto` API for crypto libraries to adapt against publicly at this time, the current proposal also avoids introducing a breaking change for enclave crypto libraries in the future.
+  - The deeper refactor to use OpenSSL directly in OE is effectively an optimization at that point by removing a layer of indirection, but at the cost of a full rewrite of the `oecryptombedtls` wrapper to continue supporting it. The relative value of doing this work is currently lower than just allowing OpenSSL to be used by the OE runtime; it already allows other OpenSSL-compatible libraries to be ported more easily into OE enclaves, while not incurring the immediate `oecryptombedtls` refactor costs.
+- OpenSSL is also currently in the process of its transition to its next major release (3.0), and it makes sense to target that version intercept as the more broadly stable crypto API surface to support.
```

Comment:
I would agree with this statement IF we have evidence that other libraries (wolfssl, libressl, etc.) will target it as well.   If the current API stays defacto standard and openssl diverges and no one follows, then that would be a bad move to pick the one that's not the defacto standard.   Today, I only have evidence that the current openssl api is a de facto standard.

Reply:
That's already kind of untrue because there are already a range of incompatible OpenSSL APIs: the move from 1.0 to 1.1 was explicitly done by OpenSSL to introduce a set of breaking API changes, and the jump from 1.1 to 3.0 is also to adopt a versioning format more aligned with what folks understand with semantic versioning instead of OpenSSL's traditional use where minor versions are API breaking. 

In general, it's possible that other distributions may choose not to follow the changes that OpenSSL makes and such choices are likely to be idiosyncratic. Beyond the API question, OE would need to move to 3.0 itself at some point, and if the argument is that it needs to support as a adapter interface it does not directly use (i.e. 1.1) then that is actually a counterargument to removing the oecrypto abstraction. (i.e. it will need to keep the current OpenSSL 1.1 to oecrypto shim for purely API reasons, but it will be using a different OpenSSL 3.0 to oecrypto shim for its actual use of the latest patched OpenSSL version.)